### PR TITLE
Please build selfbooting sdcard images #9 Sandbox/pcoval/for/master

### LINKED ISSUE #9: Please build selfbooting sdcard images
Patch to follow

### PR DESCRIPTION
**Pull request by rzr**
Fixes #9

### CODE REVIEW

**Comment by floion** [comment]
Hi @rzr , let me get this right, you want to be able to boot an SD card image and run from that, without installing to eMMC?

**Comment by rzr** [comment]
yea 100% correct

**Comment by rzr** [comment]
Is it ok to be merged ? 

FYI, I am about to wrote a tutorial about using meta-artik and would prefer to use upstream source than my fork with those patches, To be tracked at https://blogs.s-osg.org/author/pcoval/ .

Btw maybe that bug could be fixed before or after this PR , It's same to me  :-)
https://github.com/resin-os/meta-artik/issues/12

**Comment by floion** [comment]
Hi Phil, I will take a look at this PR soon, I got a little sidetracked with some other issues until now.
Will keep you posted. Thanks for submitting

File: recipes-bsp/u-boot/u-boot-artik.bb [hunk 1]
```diff
@@ -21,3 +21,33 @@ PACKAGE_ARCH = "${MACHINE_ARCH}"
...
 RDEPENDS_${PN} = "secure-boot-artik"
+
+do_compile_append() {
+    cp `find . -name "env_common.o"` copy_env_common.o
+    ${OBJCOPY} -O binary --only-section=.rodata `find . -name "copy_env_common.o"`
+    tr '\0' '\n' < copy_env_common.o | grep '=' | tee default_envs_emmc.txt default_envs_sd.txt > /dev/null
+
+    # our root partition is number 2 (exit with code 1 when no replacement done so we error out when upstream modifies the u-boot env)
+    sed -i '/rootpart=[0-9]\{1\}/{s//rootpart=2/;h};${x;/./{x;q0};x;q1}' default_envs_emmc.txt
+    sed -i '/rootpart=[0-9]\{1\}/{s//rootpart=2/;h};${x;/./{x;q0};x;q1}' default_envs_sd.txt
```
Comment (floion) [inline recipes-bsp/u-boot/u-boot-artik.bb]: Could you redo this part with a variable defined in this file, for example ROOTPART ?= "2" and use it here in this sed? This way, other layers needing to change the root partition would just redefine this ROOTPART variable in the u-boot-artik.bbappend.

File: conf/machine/include/artik.inc [hunk 2]
```diff
@@ -1,13 +1,16 @@
...
 PREFERRED_PROVIDER_u-boot ?= "u-boot-artik"
 PREFERRED_PROVIDER_virtual/kernel ?= "linux-yocto-artik"
 
 MACHINE_FEATURES += " bluetooth pci rtc ethernet serial usbgadget wifi"
 
 KERNEL_IMAGETYPE = "zImage"
+IMAGE_FSTYPES += "tar.bz2 ext4 artik-sdimg"
 
 MACHINE_ESSENTIAL_EXTRA_RRECOMMENDS += "bcm4354-bt bcm4354-nvram-config"
+MACHINE_ESSENTIAL_EXTRA_RRECOMMENDS += " kernel-modules"
```
Comment (floion) [inline conf/machine/include/artik.inc]: Also, can you remove the leading space from " kernel-modules" please? += automatically adds this space

File: classes/artik-sdimg.bbclass [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+
+	# Create boot partition and mark it as bootable
+	parted -s ${SDIMG} unit KiB mkpart primary fat32 ${IMAGE_ROOTFS_ALIGNMENT} $(expr ${BOOT_SPACE_ALIGNED} \+ ${IMAGE_ROOTFS_ALIGNMENT})
+	parted -s ${SDIMG} set 1 boot on
+
+	# Create rootfs partition to the end of disk
+	parted -s ${SDIMG} -- unit KiB mkpart primary ext2 $(expr ${BOOT_SPACE_ALIGNED} \+ ${IMAGE_ROOTFS_ALIGNMENT}) -1s
+
+	# Set boot flag
+	parted -s ${SDIMG} set 1 boot on
```
Comment (floion) [inline classes/artik-sdimg.bbclass]: This has been already done on line 92

File: classes/artik-sdimg.bbclass [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+	rm -f ${WORKDIR}/boot.img
+	mkfs.vfat -n "${BOOTDD_VOLUME_ID}" -S 512 -C ${WORKDIR}/boot.img $BOOT_BLOCKS
+
+
+	mcopy -i ${WORKDIR}/boot.img -s ${DEPLOY_DIR_IMAGE}/${KERNEL_IMAGETYPE} ::${KERNEL_IMAGETYPE}
+	if test -n "${KERNEL_DEVICETREE}"; then
+		mcopy -i ${WORKDIR}/boot.img -s ${DEPLOY_DIR_IMAGE}/${KERNEL_IMAGETYPE}-${KERNEL_DEVICETREE} ::${KERNEL_DEVICETREE}
+	fi
+
+	if [ -n ${FATPAYLOAD} ] ; then
```
Comment (floion) [inline classes/artik-sdimg.bbclass]: Maybe also define FATPAYLOAD with a null string along with a short comment on what it should be used, just above the IMAGE_CMD_artik-sdimg () function start

**Comment by rzr** [comment]
Thanks

**Comment by agherzan** [comment]
Good work!

**Pull request merged**